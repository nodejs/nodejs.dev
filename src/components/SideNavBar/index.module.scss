@import '../../containers/Navigation/index.module.scss';
@import '../NavigationSection/index.module.scss';
@import '../NavigationItem/index.module.scss';

.side-nav {
  @extend %navigation;

  &__open {
    @extend %navigation__open;
  }

  &--fixed {
    @extend %navigation--fixed;

    height: 100vh;
    top: calc(var(--nav-height) * 2);
    touch-action: none;
    width: 100%;
  }

  &--fixed &__open {
    top: 1.8rem;
  }

  .side-nav-section {
    @extend %navigation-section;

    padding: 0 var(--space-20);

    &__title {
      @extend %navigation-section__pure-title;

      margin-left: 0;

      b {
        color: var(--color-brand-primary);
        font-weight: var(--font-weight-semibold);
      }
    }

    .side-nav-item {
      @extend %navigation-item;

      padding: calc(var(--space-20) / 2) 0 calc(var(--space-20) / 2)
        var(--space-12);

      &::before {
        box-shadow: inset 0 0 0 2px var(--color-border-accent);
        line-height: 2.6rem;
        top: 1rem;
      }

      &::after {
        content: none;
      }

      &:hover::before {
        box-shadow: inset 0 0 0 2.4rem var(--color-border-accent) !important;
      }

      &--active {
        color: var(--color-text-accent);
        font-weight: 600;

        &::before {
          box-shadow: inset 0 0 0 2.4rem var(--color-border-accent);
        }
      }
    }
  }
}
