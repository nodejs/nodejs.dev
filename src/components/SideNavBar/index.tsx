import React, { useRef, useState, useMemo, useEffect } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import classnames from 'classnames';
import { SideNavBarItem } from '../../types';
import NavigationItem from '../NavigationItem';
import styles from '../../styles/navigation.module.scss';

// eslint-disable-next-line no-shadow
export enum SideNavBarKeys {
  about = 'about',
  governance = 'about/governance',
  community = 'community',
  workingGroups = 'about/working-groups',
  releases = 'about/releases',
  resources = 'resources',
  privacy = 'about/privacy',
  security = 'about/security',
  packageManager = 'download/package-manager',
}

const sideNavBarItems: SideNavBarItem[] = [
  {
    title: 'components.sideBar.items.about',
    slug: SideNavBarKeys.about,
  },
  {
    title: 'components.sideBar.items.governance',
    slug: SideNavBarKeys.governance,
  },
  {
    title: 'components.sideBar.items.community',
    slug: SideNavBarKeys.community,
  },
  {
    title: 'components.sideBar.items.workingGroups',
    slug: SideNavBarKeys.workingGroups,
  },
  {
    title: 'components.sideBar.items.releases',
    slug: SideNavBarKeys.releases,
  },
  {
    title: 'components.sideBar.items.resources',
    slug: SideNavBarKeys.resources,
  },
  {
    title: 'components.sideBar.items.privacy',
    slug: SideNavBarKeys.privacy,
  },
  {
    title: 'components.sideBar.items.security',
    slug: SideNavBarKeys.security,
  },
  {
    title: 'components.sideBar.items.packageManager',
    slug: SideNavBarKeys.packageManager,
  },
];

// eslint-disable-next-line no-shadow
export enum OverflowTypes {
  unset = 'unset',
  hidden = 'hidden',
}

interface NavBarProps {
  pageKey?: string;
  items?: SideNavBarItem[];
  title?: string;
}

const SideNavBar = ({
  pageKey,
  items = sideNavBarItems,
  title,
}: NavBarProps): JSX.Element => {
  const [navOpen, setNavOpen] = useState<boolean>(false);
  const toggle = (): void => setNavOpen(!navOpen);
  const intl = useIntl();

  const navigationClasses = classnames(
    styles.sideNav,
    styles.sideNavCommunity,
    { [styles.sideNavFixed]: navOpen, [styles.sideNavCommunityFixed]: navOpen }
  );

  useEffect(() => {
    if (typeof document === 'object' && document.body) {
      document.body.style.overflow = navOpen
        ? OverflowTypes.hidden
        : OverflowTypes.unset;
    }
  }, [navOpen]);

  const navElement = useRef<HTMLElement | null>(null);

  const renderNavTitle = () => {
    if (title) {
      return (
        <li className={`t-body2 ${styles.sideNavListItemTitle}`}>
          <b>{title}</b>
          <i className="material-icons">arrow_drop_down</i>
        </li>
      );
    }

    return null;
  };

  const translatedSidebar = useMemo(
    () =>
      items.map(item => ({
        ...item,
        // Not all entries necessarily need to be translated
        // As some are non-translatable bags
        title: intl.messages[item.title]
          ? intl.formatMessage({ id: item.title })
          : item.title,
      })),
    [intl, items]
  );

  return (
    <nav className={navigationClasses} ref={navElement}>
      <button type="button" className={styles.sideNavOpen} onClick={toggle}>
        <FormattedMessage id="components.sideBar.title" />
      </button>
      <div className={styles.sideNavList}>
        {renderNavTitle()}
        {translatedSidebar
          .sort((a, b) => a.title.localeCompare(b.title))
          .map(({ title: commTitle, slug }) => (
            <NavigationItem
              key={slug}
              title={commTitle}
              isRead={false}
              isActive={slug === pageKey}
              slug={slug}
              baseUrl="/"
            />
          ))}
      </div>
    </nav>
  );
};

export default SideNavBar;
