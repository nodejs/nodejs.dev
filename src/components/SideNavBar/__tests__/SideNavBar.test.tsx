import React from 'react';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import SideNavBar, { SideNavBarKeys, OverflowTypes } from '..';

describe('SideNavBar', () => {
  it('renders correctly', () => {
    const { container } = render(
      <SideNavBar pageKey={SideNavBarKeys.trademark} />
    );
    expect(container).toMatchSnapshot();
  });

  it('should set the a single page as active', async () => {
    const { container } = render(
      <SideNavBar pageKey={SideNavBarKeys.releases} />
    );
    const innrerHtml = container.innerHTML;
    const activeLinks = innrerHtml.match('side-nav__item-community--active');
    expect(activeLinks && activeLinks.length).toBe(1);
  });

  it('should contain a href to `~/about`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.about} />);
    const aboutNavBarElement = screen.getByText('About');
    expect(aboutNavBarElement.getAttribute('href')).toBe('/about/');
  });

  it('should contain a href to `~/about/governance`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.governance} />);
    const governanceNavBarElement = screen.getByText('Project Governance');
    expect(governanceNavBarElement.getAttribute('href')).toBe(
      '/about/governance/'
    );
  });

  it('should contain a href to `~/community`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.community} />);
    const communityNavBarElement = screen.getByText('Community');
    expect(communityNavBarElement.getAttribute('href')).toBe('/community/');
  });

  it('should contain a href to `~/about/working-groups`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.workingGroups} />);
    const workingGroupsNavBarElement = screen.getByText('Working Groups');
    expect(workingGroupsNavBarElement.getAttribute('href')).toBe(
      '/about/working-groups/'
    );
  });

  it('should contain a href to `~/about/releases`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.releases} />);
    const releasesNavBarElement = screen.getByText('Releases');
    expect(releasesNavBarElement.getAttribute('href')).toBe('/about/releases/');
  });

  it('should contain a href to `~/about/resources`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.resources} />);
    const resourcesNavBarElement = screen.getByText('Resources');
    expect(resourcesNavBarElement.getAttribute('href')).toBe('/resources/');
  });

  it('should contain a href to `~/about/trademark`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.trademark} />);
    const trademarkNavBarElement = screen.getByText('Trademark Policy');
    expect(trademarkNavBarElement.getAttribute('href')).toBe(
      '/about/trademark/'
    );
  });

  it('should contain a href to `~/about/privacy`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.privacy} />);
    const privacyNavBarElement = screen.getByText('Privacy Policy');
    expect(privacyNavBarElement.getAttribute('href')).toBe('/about/privacy/');
  });

  it('should contain a href to `~/about/security`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.security} />);
    const securityNavBarElement = screen.getByText('Security Reporting');
    expect(securityNavBarElement.getAttribute('href')).toBe('/about/security/');
  });

  it('should contain a href to `~/download/package-manager`', () => {
    render(<SideNavBar pageKey={SideNavBarKeys.packageManager} />);
    const packageManagerNavBarElement = screen.getByText('Package Manager');
    expect(packageManagerNavBarElement.getAttribute('href')).toBe(
      '/download/package-manager/'
    );
  });

  it('should set the a single page as active', () => {
    const { container } = render(
      <SideNavBar pageKey={SideNavBarKeys.releases} />
    );
    const innrerHtml = container.innerHTML;
    const activeLinks = innrerHtml.match('side-nav__item-community--active');
    expect(activeLinks && activeLinks.length).toBe(1);
  });

  it('should set the body overflow hidden on menu click', async () => {
    render(<SideNavBar pageKey={SideNavBarKeys.releases} />);
    const downloadItem: Element = screen.getAllByRole('button')[0] as Element;
    expect(document.body.style.overflow).toBe(OverflowTypes.unset);
    await userEvent.click(downloadItem);
    expect(document.body.style.overflow).toBe(OverflowTypes.hidden);
  });

  it('should set the body overflow hidden/unset on toggling', async () => {
    render(<SideNavBar pageKey={SideNavBarKeys.releases} />);
    const downloadItem: Element = screen.getAllByRole('button')[0] as Element;
    expect(document.body.style.overflow).toBe(OverflowTypes.unset);
    await userEvent.click(downloadItem);
    expect(document.body.style.overflow).toBe(OverflowTypes.hidden);
    await userEvent.click(downloadItem);
    expect(document.body.style.overflow).toBe(OverflowTypes.unset);
    await userEvent.click(downloadItem);
    expect(document.body.style.overflow).toBe(OverflowTypes.hidden);
    await userEvent.click(downloadItem);
    expect(document.body.style.overflow).toBe(OverflowTypes.unset);
  });
});
