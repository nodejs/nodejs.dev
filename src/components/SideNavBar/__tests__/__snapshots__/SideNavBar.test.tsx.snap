// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavBar renders correctly 1`] = `
<div>
  <nav
    class="sideNav"
  >
    <button
      class="sideNavOpen"
      type="button"
    >
      Menu
    </button>
    <div
      class="sideNavListCommunity"
    >
      <a
        aria-current="page"
        class="t-body2 sideNavListItem sideNavListItemActive"
        href="/about/"
        id="link-about"
      >
        About
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/community/"
        id="link-community"
      >
        Community
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/download/package-manager/"
        id="link-download/package-manager"
      >
        Package Manager
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/about/privacy/"
        id="link-about/privacy"
      >
        Privacy Policy
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/about/governance/"
        id="link-about/governance"
      >
        Project Governance
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/about/releases/"
        id="link-about/releases"
      >
        Releases
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/resources/"
        id="link-resources"
      >
        Resources
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/about/security/"
        id="link-about/security"
      >
        Security Reporting
      </a>
      <a
        class="t-body2 sideNavListItem"
        href="/about/working-groups/"
        id="link-about/working-groups"
      >
        Working Groups
      </a>
    </div>
  </nav>
</div>
`;
